A first draft of "loading data: images"

MarkDaoust opened the pull request:
It still needs an introduction and conclusion paragraph.

But this won't be visible on the site unless we add it to the "_toc.yaml"

Staging:

https://github.com/MarkDaoust/docs/blob/load_data/site/en/tutorials/load_data/images.ipynb

https://colab.sandbox.google.com/github/MarkDaoust/docs/blob/load_data/site/en/tutorials/load_data/images.ipynb

random-forests commented:
Thank you for writing this! Good timing, too. Together with Kiran's notebook on tf records, this would make a great start to a "Reading data" collection.

Suggestions:
=========

Under "Load and format the images"

* Instead of normalizing the image by "image = (image/128.0) - 1", instead, how about "image = image / 255"? 

* This would also simplify the code to display the images (instead of "plt.imshow((image+1)/2)", we can write "plt.imshow(image)")

* How about demo'ing the "load_and_format_image" function, just to demonstrate it works like regular Python code when eager is enabled? To do so, you could add a short block under it, that calls the method using something like "load_and_format_image('/root/.keras/datasets/flower_photos/tulips/7166644048_b00a14f01b.jpg', 0)" (I know there's a bunch of output, maybe catch it in a object, show how to get the .numpy() value, and print out the shape

Under "Pipe to a model for training"

* Explain the prefetch parameter. 

Under "Quick Transfer learning with keras.Applications"

* Leave the depth_multiplier at the default value.

* Finally, split the dataset into train / test, so we can show a quick evaluation after the transfer learning section, which is really cool. 

Overall this is great!

MarkDaoust commented:
> Instead of normalizing the image by "image = (image/128.0) - 1", instead, how about "image = image / 255"?

Done.
The `tf.keras.applications.mobilenet` expects data to be normalized to [-1,1].
I've switched it to use [0,1] for most of it, with a transform to [-1,1] in the transfer learning section. 

>demo'ing the "load_and_format_image"

Done.

>Explain the prefetch parameter.

Done

>Leave the depth_multiplier at the default value.

Done

>Finally, split the dataset into train / test, so we can show a quick evaluation after the transfer learning section, which is really cool.

I'd rather avoid going into too much detail here. I just want to get to "see how easy that was!" and leave the rest for a real tutorial.

MarkDaoust commented:
+tf.data

Hey Derek, Jiri, 

Does this seem like a reasonable approach to loading image directories with `tf.data`?

jsimsa commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "code",
+      "source": [
+        "image_ds = path_ds.map(load_and_preprocess_image, num_parallel_calls=8)"
```
replace `num_parallel_calls=8` with `num_parallel_calls=tf.data.experimental.AUTOTUNE`

jsimsa commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "ds = ds.shuffle(buffer_size=4000) \n",
+        "ds = ds.batch(BATCH_SIZE)\n",
+        "# `prefecth` lets the dataset fetches batches, asynchronously while the model is training.\n",
+        "ds = ds.prefetch(1)\n",
```
replace `1` with `buffer_size=tf.data.experimental.AUTOTUNE`

mrry replied:
+1. I wonder if it would make sense to add this step inside Keras/Distribution Strategies as well?

jsimsa commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      "source": [
+        "def change_range(image,label):\n",
+        "  image = 2*image-1\n",
+        "  return image, label\n",
```
this could just be `return 2*image - 1, label`

jsimsa commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      "cell_type": "code",
+      "source": [
+        "ds = image_label_ds.repeat()\n",
+        "ds = ds.shuffle(buffer_size=4000) \n",
```
OOC, why did you pick 4000? as someone following the tutorial, I would want to know where this magic constant comes from. It might be worth saying that smaller values will result in faster startup (the buffer to initial fill is smaller) while higher values result in better shuffling (in the limit where buffer_size == input_size the shuffling is perfectly random).

Another thing to note is that because you `repeat` before `shuffle` some images might occur for the second time before other images occur at all. This can generally negatively affect accuracy. I suspect the reason you are doing this is to avoid filling up the shuffle buffer at the beginning of each epoch. To avoid doing that, while maintaining clean epoch boundaries, you can either 1) use the [tf.data.experimental.shuffle_and_repeat](https://www.tensorflow.org/api_docs/python/tf/data/experimental/shuffle_and_repeat) transformation or 2) do this:

```
ds = ds.shuffle(...).repeat().
options = tf.data.Options
options.experimental_shuffle_and_repeat_fusion = True
ds = ds.with_options(ds)
```

MarkDaoust replied:
hmmm... the options approach doesn't seem to be working in tf-nightly. I'll send you a bug.

`shuffle_and_repeat` is.
Fixed.

jsimsa replied:
The API for this has changed to:

```
ds = ds.shuffle(...).repeat().
options = tf.data.Options
options.experimental_optimization = tf.data.experimental.OptimizationOptions()
options.experimental_optimization.shuffle_and_repeat_fusion = True
ds = ds.with_options(ds)
```

jsimsa commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "\n",
+        "Note: This section just shows a couple of easy tricks that may help performance. For an in depth guide see [Input Pipeline Performance](https://www.tensorflow.org/guide/performance/datasets)\n",
+        "\n",
+        "The simple pipeline used above reads each file individually, on each epoch. This is fine for local training on CPU but may not be sufficient for GPU training, and is totally inapprpriate for any sort of distributed training. "
```
fix typo in `inapprpriate`

jsimsa commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "markdown",
+      "source": [
+        "Use `tf.data.Dataset.cache` to easily take advantage of the performance boost of working with in memory data.\n",
```
Make note of the fact that `cache` is only beneficial if the input dataset fits into memory. If it does not, then using `cache` does more harm then good.

jsimsa replied:
nevermind, you already talk about this later

jsimsa commented:
Thanks @MarkDaoust did a pass and left some comments.

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "markdown",
+      "source": [
+        "The `output_shapes` and `output_types` fields describe the content of each item in the dataset. In this case it is a set of scalar binary-strings"
```
We're going to remove these properties in V2. (They were rarely used by user code, and they aren't sufficient to express all of the things that can be in a dataset... especially the nested output of `Dataset.window()`.) The replacement will most likely be a single `Dataset.element_structure` property (exact name TBD) that serves the same purpose, and provides a superset of the same information.

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "markdown",
+      "source": [
+        "Since the datasets are in the same order we can just zip them together to get a dataset of `(image, label)` pairs."
```
Would it be worth mentioning that you can also call `from_tensor_slices()` with a pair of tensors directly?

MarkDaoust replied:
I only did it this way because we had the images in a Dataset already.

I've added an example to the end of this section demonstrating:

```
ds = tf.data.Dataset.from_tensor_slices((all_image_paths, all_image_labels))

# The tuple elements are unpacked into the positional arguments of the mapped function 
def load_and_preprocess_from_path_label(path, label):
  return load_and_preprocess_image(path), label

image_label_ds = ds.map(load_and_preprocess_from_path_label)
image_label_ds
```

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "markdown",
+      "source": [
+        "The new dataset's `shapes` and `types`, are tuples of shapes and types as well, describing each field:"
```
Nit: no comma after `types`.

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      "source": [
+        "To train a model with this dataset you will want the data:\n",
+        "\n",
+        "* To be well shuffeled.\n",
```
s/shuffeled/shuffled/

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "\n",
+        "* To be well shuffeled.\n",
+        "* To be batched.\n",
+        "* To repeat forever.\n",
```
This part is TBD with the Keras API. I think we're leaning towards *not* repeating the dataset itself forever, but instead iterating over the same finite dataset for a given number of epochs inside Keras.

MarkDaoust replied:
Acknowledged.

+1, that would be  a more intuitive API (I've already mentioned this to the keras team in b/109665994. Do you know the status of that plan?).

AFAIK, currently the only way it works is for the user to pass a `Dataset` that's long enough (typically using `.repeat()`) and use the `steps_per_epoch` argument to `fit`.

So I'll leave this for now, and gladly fix this when the behavior is changed.

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "* To be well shuffeled.\n",
+        "* To be batched.\n",
+        "* To repeat forever.\n",
+        "* Batches to available immediately when needed.\n",
```
"to be available as soon as possible"?

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "ds = image_label_ds.repeat()\n",
+        "ds = ds.shuffle(buffer_size=4000) \n",
+        "ds = ds.batch(BATCH_SIZE)\n",
+        "# `prefecth` lets the dataset fetches batches, asynchronously while the model is training.\n",
```
s/fetches batches, asynchronously/fetch batches asynchronously/

mrry replied:
s/prefecth/prefetch/

mrry replied:
Actually, for a novice user, I wonder if "in the background" would be more intuitive than "asynchronously"?

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "markdown",
+      "source": [
+        "This model expects it's input to be normalized to the `[-1,1]` range:\n",
```
s/it's/its/

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "markdown",
+      "source": [
+        "So before the passing it to the mobilnet model, we need to convert the input from a range of `[0,1]` to `[-1,1]`."
```
s/mobilnet/MobileNet/ (or `mobile_net`)

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      "cell_type": "code",
+      "source": [
+        "# The dataset may take a few seconds to start, as it fills its shuffle buffer.\n",
+        "image_batch, label_batch = next(iter(keras_ds))"
```
@jsimsa I forget: did we discuss adding a `Dataset.first()` method to make this a bit cleaner? (IIRC it was going to make some of the reducer examples a lot nicer.)

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "code",
+      "source": [
+        "model.fit(ds, epochs=1, steps_per_epoch=3)  # steps_per_epoch=math.ceil(len(all_image_paths)/BATCH_SIZE)"
```
It doesn't exist yet, but I think the 2.0 recommendation for this will be `model.fit(ds.take(3), epochs=1)` (when the dataset input is finite).

MarkDaoust replied:
Great. LMK when that happens and I can scrub `steps_per_epoch` from our docs.

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "markdown",
+      "source": [
+        "One disadvantage to using an im memory cache is that the cache must be rebuilt on each run, giving the same startup delay each time the dataset is started:"
```
I think @rohan100jain has some changes in flight that will help this no longer be the case.

MarkDaoust replied:
!

How does that work?

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "markdown",
+      "source": [
+        "TFRecord files are a simple format to store a sequence of binary blobs. In this case the fact that these are read linearly off disk, which may give some performance boost.\n",
```
The second sentence doesn't make much sense to me. Perhaps something like "By packing multiple examples into the same file, TensorFlow is able to read multiple examples at once, which is especially important for performance when using a remote storage service such as GCS."?

MarkDaoust replied:
SGTM

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      "source": [
+        "TFRecord files are a simple format to store a sequence of binary blobs. In this case the fact that these are read linearly off disk, which may give some performance boost.\n",
+        "\n",
+        "It also gives you a portable data file that is apropriate, for example, for streaming over a network.\n",
```
s/apropriate/appropriate/

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "\n",
+        "It also gives you a portable data file that is apropriate, for example, for streaming over a network.\n",
+        "\n",
+        "First, build a `TFRecord` file from the raw image data:"
```
Nit: "TFRecord" isn't usually surrounded by backticks.

mrry commented on site/en/tutorials/load_data/images.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "cell_type": "markdown",
+      "source": [
+        "This is slower than the `cache` version because we have not cached the preprocessing."
```
This is kind of an anticlimax! Any chance of a "However, ..."?

MarkDaoust replied:
Sure.

I've split the TFRecord section into:

1. "Raw image data", where I dump the jpeg data directly to the TFRecord file.
2. "Serialized Tensors", where I decode and resize the images and `serialize_tensor` them into the TFRecord file.

MarkDaoust commented:
(not really merging, just testing copybara)
